### PR TITLE
Implement canary ingresses

### PR DESCRIPTION
**Pull request by clandry94**
Implements the feature described here https://github.com/Shopify/edgescale/issues/664 

This PR adds the initial support for multi-service routing per ingress by means of alternative backends. It also implements an interface to create these alternative backends through the form of canary deploys. 

### Some terms
**AlternativeBackend** - An alternative backend is a normal ingress-nginx backend with the condition that it is not backed through the typical 1:1 backend-server relationship. Instead, alternative backends share a server with another real backend.

**Primary Backend** - This is the same as a backend. I prefix with the word primary to have a clearer distinction between it and a alternative backend

#### Canary
As described in the linked issue above, the following annotation
```
nginx.ingress.kubernetes.io/canary: true
```
Combined with a combination of the below annotations, we can define the behaviour of the canary. 

The canary will define a `canary-weight`. If only weight is specified, then n% of traffic will be shifted to the canary backend. Additionally, if `canary-by-cookie` or `canary-by-header` are set, they will shift all traffic which match the cookie/header set with a value of `always` to the canary backend. If no value is set for the header, it will follow the rules of `canary-weight`. 

```
nginx.ingress.kubernetes.io/canary-by-cookie: <cookie name>
nginx.ingress.kubernetes.io/canary-by-header: <header name>
nginx.ingress.kubernetes.io/canary-weight: <0 to 100>
```

After the spec is applied, the ingress in question has it's backends marked as alternative so that the merging code will know what to do with them.

#### AlternativeBackends

When ingresses are synced, the list of backends and servers is built up. After these lists are finished being built in `getBackendServers()`, but before they are sorted and returned to the sync function, we check if the current ingress has a canary enabled. If so, we pass a reference of the ingress, annotations, upstreams, and servers to the `mergeVirtualBackends` function. It works as such:
1) For each path that matches a alternative backend, find the server that matches its path and is not alternative. 
2) If found, embed a alternative backend into the primary backend. If no matching path is found, then delete the alternative upstream as it is invalid and has no server to back it. 

When alternative backends are matched to a primary backend, they will have access to metadata that describes the primary backend as well as rules to shape traffic. These rules come from the canary ingress configuration.

### CODE REVIEW

**Comment by clandry94** [comment]
General question I have that I don't like my current solution for (not implemented yet):

At the moment, dynamic config seems to mean "changes only occurred in the backends" as shown by this method: https://github.com/Shopify/ingress/blob/3019560e28312923600d1e7844e10dd84349b823/internal/ingress/controller/nginx.go#L740-L755

The nature of virtual backends means they will fall under the dynamic configuration (no servers are modified). As a result, I'm not sure how we will get new information regarding virtual backends into the nginx configuration. Should I force a backend reload if a virtual ingress is detected on a backend?

File: internal/ingress/annotations/canary/main_test.go
Comment (zrdaley) [inline internal/ingress/annotations/canary/main_test.go]: Is there a reason why there isn't a check for `canaryHeader` and `canaryCookie`?

File: internal/ingress/controller/controller.go
Comment (zrdaley) [inline internal/ingress/controller/controller.go]: Should this be initialized to false? Otherwise it will always be true.

File: internal/ingress/annotations/canary/main_test.go
Comment (clandry94) [reply internal/ingress/annotations/canary/main_test.go]: No, I just forgot to add those checks. I'll add them

File: internal/ingress/controller/controller.go
Comment (zrdaley) [inline internal/ingress/controller/controller.go]: You only call mergeVirtualBackends when anns.Canary.Enabled is true. Are these markings necessary? It seems like a redundant check.

File: internal/ingress/controller/controller.go
Comment (Wayt) [inline internal/ingress/controller/controller.go]: Why not:
```
upstreams[name].Virtual = anns.Canary.Enabled
```

File: internal/ingress/annotations/canary/main_test.go
Comment (Wayt) [inline internal/ingress/annotations/canary/main_test.go]: What if you don't expect and error and you have one ?

File: internal/ingress/controller/controller.go
Comment (clandry94) [reply internal/ingress/controller/controller.go]: Setting the `Virtual` field to true on the backend is necessary since its used in the merge function here:
https://github.com/Shopify/ingress/blob/abd198fc93825399648a659bccec6b77ecd3cc7b/internal/ingress/controller/controller.go#L1142

My motivation here was to not couple the idea of canary too tightly with concept of virtual backends to stay flexible at the cost of some redundancy.

File: internal/ingress/controller/template/template.go
Comment (Wayt) [inline internal/ingress/controller/template/template.go]: why ?

File: internal/ingress/annotations/canary/main_test.go
Comment (clandry94) [reply internal/ingress/annotations/canary/main_test.go]: Updated to handle that case now
https://github.com/Shopify/ingress/blob/97a1ebe97cd7e43e61da0fd7c5dea6f1e2520a56/internal/ingress/annotations/canary/main_test.go#L96-L107

File: internal/ingress/types_equals.go
Comment (DazWorrall) [inline internal/ingress/types_equals.go]: Should this be `vb1.Cookie != vb2.Cookie`?

**Review by wayt — DISMISSED** [review]

File: internal/ingress/annotations/canary/main.go
Comment (ElvinEfendi) [inline internal/ingress/annotations/canary/main.go]: nitpick: instead of this comment I'd rewrite the error message to actually say what's the error.

File: internal/ingress/annotations/canary/main_test.go
Comment (ElvinEfendi) [inline internal/ingress/annotations/canary/main_test.go]: nitpick: `i`, `u` does not tell me anything about the object it's referring

File: internal/ingress/controller/controller.go
Comment (ElvinEfendi) [inline internal/ingress/controller/controller.go]: can you add more description here explaining how it matches, how it merges. Ideally with examples

File: internal/ingress/controller/controller.go
Comment (ElvinEfendi) [inline internal/ingress/controller/controller.go]: could you create a utility function for this and DRY it everywhere else in this package as well?

File: internal/ingress/controller/controller.go
Comment (ElvinEfendi) [inline internal/ingress/controller/controller.go]: If I read this correctly you are matching by path only and ignoring server name (host). That means given two apps with same path `/` and different domains respectively myapp.com, yourapp.com and I create canary ingress for my app it will also get applied to yourapp.com

File: internal/ingress/controller/controller_test.go
Comment (ElvinEfendi) [inline internal/ingress/controller/controller_test.go]: can you create another server here where the hostname is foo.bar and assert that virtual backend won't match it

File: internal/ingress/types.go
Comment (ElvinEfendi) [inline internal/ingress/types.go]: is this Go convention to have an extra line here?

File: internal/ingress/types.go
Comment (ElvinEfendi) [inline internal/ingress/types.go]: please document what's virtual backend for, and how it's used

File: internal/ingress/types.go
Comment (clandry94) [reply internal/ingress/types.go]: I don't remember adding this line, maybe gofmt did? I'll remove it and see if any checks yell at me

File: internal/ingress/types.go
Comment (ElvinEfendi) [inline internal/ingress/types.go]: I find this name counterintuitive. This is not a backend, it is rather a construct to augment actual backend object and make it "virtual".

Comment (ElvinEfendi) [reply internal/ingress/types.go]: A virtual backend is actually of type `Backend` where `Virtual` attribute is `true`.

File: internal/ingress/annotations/canary/main.go
Comment (ElvinEfendi) [inline internal/ingress/annotations/canary/main.go]: please check `internal/ingress/errors/errors.go` and see if it makes more sense to use that package for better error handling

File: rootfs/etc/nginx/lua/balancer.lua
Comment (ElvinEfendi) [inline rootfs/etc/nginx/lua/balancer.lua]: we should not do json decoding in-line with request

**Comment by ElvinEfendi** [comment]
Great work @clandry94 ! I am not convinced that naming is clear enough though. `VirtualBackend` does not represent what it is for at all IMO.

Also we will have to refactor balancer.lua to accommodate this functionality, what you've done in Lua side to support this routing seems hacky - we should avoid decoding JSON in-line with request to avoid extra processing. One option is when generating balancers we can extract `VirtualBackend`s and store them per backend name. Something like
```
{
  <my backend>: {
    <virtual backend 1>: { <virtual backend setting> },
    ....
  }
}
then we can access VirtualBackend's for given request easily and make routing decision.
```

**Comment by clandry94** [comment]
@ElvinEfendi Thanks! I wasn't happy with the json decoding inline with requests, but I wanted to get your take on it first. I'll look into storing them per backend when generating the balancer

File: Makefile
Comment (zrdaley) [inline Makefile]: Also don't forget to remove this before merging.

File: internal/ingress/types.go
Comment (clandry94) [reply internal/ingress/types.go]: Nice. I would use goimports, but CI for ingress-nginx requires gofmt to be run

File: internal/ingress/annotations/canary/main_test.go
Comment (clandry94) [inline internal/ingress/annotations/canary/main_test.go]: I left this as `i` since its an intermediary between it and L109 where its type casted to a `canary.Config`

File: internal/ingress/controller/controller.go
Comment (clandry94) [reply internal/ingress/controller/controller.go]: I've added a check for host a few lines above this now.

File: internal/ingress/controller/controller.go
Comment (clandry94) [reply internal/ingress/controller/controller.go]: Now that the virtual backend metadata is set when this check happens, the if statement is needed.

File: internal/ingress/controller/controller_test.go
Comment (clandry94) [reply internal/ingress/controller/controller_test.go]: Added that test to ensure a 1:1 match by host name

File: Makefile
Comment (clandry94) [inline Makefile]: Will revert remove before merging

File: .travis.yml
Comment (clandry94) [inline .travis.yml]: Will revert before merging

File: internal/ingress/types.go
Comment (clandry94) [inline internal/ingress/types.go]: @ElvinEfendi How is this rather than the `VirtualBackend` struct? I think this is better since it also means the whole backend is initialized in the `createUpstreams` function rather than initializing virtual backends in the `mergeVirtualBackends` function.

I agree in thinking using the name `VirtualBackend` isn't very appropriate.  A true `Virtual` backend would be better represented as the top level backend being virtual and it contains a list of real backends. Perhaps something like:

```
type Backend struct {
  // Whether or not the backend has a server
  HasServer bool
  TrafficShapingConfig TrafficShapingConfig
  // List of backends that requests could also be routed to
  AlternativeBackends []*Backend
}

// Contains rules for how to route traffic to alternative backends
type TrafficShapingConfig struct {
  Weight int  
  Header string
  Cookie string
}
```

This drops the virtual connotations while being clear that a backend can standalone or not. I think a better name could be come created than `TrafficShapingConfig`, though 🤔

File: internal/ingress/annotations/canary/main_test.go
Comment (zrdaley) [inline internal/ingress/annotations/canary/main_test.go]: What if canary is enabled by both header and cookie?

File: internal/ingress/annotations/canary/main_test.go
Comment (zrdaley) [inline internal/ingress/annotations/canary/main_test.go]: This is a bit of a nitpick but it would be nice to see the error message that should be returned. 

Maybe `expErr` could be the error message string rather than a bool to make for easier debugging in the case that the test fails.

Comment (clandry94) [reply internal/ingress/annotations/canary/main_test.go]: I think that would be better as well, I just copied this format from another annotation test file. I can improve this with another PR after we merge this.

File: internal/ingress/types_equals.go
Comment (wayt) [inline internal/ingress/types_equals.go]: duplicate from line 87?

File: internal/ingress/controller/controller_test.go
Comment (zrdaley) [inline internal/ingress/controller/controller_test.go]: How does checking only the length ensure that the test `merging a virtual backend matches with the correct host` actually works as intended. Shouldn't the actual backend/location values be compared as well?

File: rootfs/etc/nginx/lua/balancer.lua
Comment (wayt) [inline rootfs/etc/nginx/lua/balancer.lua]: you're doing two lookup in this table by doing this, you could simply:
```
local header = ngx.var["http_" .. virtual_backend.virtualMetadata.header]
if header then
[...]
```

File: rootfs/etc/nginx/lua/balancer.lua
Comment (wayt) [inline rootfs/etc/nginx/lua/balancer.lua]: Same as `header`, you're doing two lookup

File: internal/ingress/types.go
Comment (wayt) [reply internal/ingress/types.go]: I too have difficulties with the `Virtual` naming.

What you're trying to do is splitting routing into multiple backends, which allows us then to do some Canary. But we should not limit it to the Canary feature.

Naming is hard, but I prefer this option over the `Virtual` one.

**Review by wayt — COMMENTED** [review]
This looks very good :) That's a great job.

I'll let the final decision/review to @ElvinEfendi and @zrdaley they have more context than me on ingress-nginx codebase.

File: rootfs/etc/nginx/lua/balancer.lua
Comment (zrdaley) [inline rootfs/etc/nginx/lua/balancer.lua]: nitpick:
```
if <condition> then
  return <whatever>
end
```

There are more cases of this below.

File: internal/ingress/controller/controller_test.go
Comment (clandry94) [reply internal/ingress/controller/controller_test.go]: My reasoning here was pretty much if the example.com host has 1 virtual backend, then the matching worked properly. 2 would mean that the `foo.bar` virtual went to the wrong real backend and 0 would mean that the `example.com` host went to the wrong real backend. Of course, this doesn't check for correctness entirely since they could be assigned in reverse order like you said. If you're okay with it, I'll improve this test when I improve the other test as well.

File: internal/ingress/controller/controller.go
Comment (ElvinEfendi) [inline internal/ingress/controller/controller.go]: Is this evey going to be `false`? Since we call this function iff the ingress is annotated with `canary`, the corresponding upstreams will all have `NoServer` set to `true`, right?

**Comment by ElvinEfendi** [comment]
@clandry94 please add e2e tests at `test/e2e/annotations/` for at least canary by header and cookie (since they are predictable).

File: internal/ingress/types.go
Comment (ElvinEfendi) [inline internal/ingress/types.go]: There should be a way to avoid sending the whole backend structs here, they are already being posted as normal backends to Lua land. Only their names should be needed here.

File: internal/ingress/controller/controller.go
Comment (clandry94) [reply internal/ingress/controller/controller.go]: There was a point before I refactored where it could have been been false, but that shouldn't be the case anymore. I can probably remove `NoServer` altogether

File: internal/ingress/types.go
Comment (clandry94) [reply internal/ingress/types.go]: I can make that work. Since every backend has a balancer in lua land, keeping just their names should be straightforward to change.

Comment (ElvinEfendi) [reply internal/ingress/types.go]: The tricky part is that you won't have access to traffic shaping policies then. I guess that can be part of the respective balancer instance. Then in Lua code instead of `alternative_backend` you can work with `alternative_balancer`.

File: internal/ingress/types.go
Comment (ElvinEfendi) [inline internal/ingress/types.go]: why are these not consistent (NoServer vs hasServer)?

**Comment by ElvinEfendi** [comment]
@clandry94 please rebase with master and change e2e tests according to https://github.com/kubernetes/ingress-nginx/pull/3325/files.

File: rootfs/etc/nginx/lua/balancer.lua
Comment (ElvinEfendi) [inline rootfs/etc/nginx/lua/balancer.lua]: This seems incorrect to me - when I set weight to `0`, that should mean no traffic to alternative backend. However with this implementation there 1/101 chance that it will send traffic to the canary backend.

I think it should be `math.random(1, 100)`.

--

> this function has a left-hand bias which produces

Are you referring to `math.random`? Can you explain more where does that bias come from?

**Comment by ElvinEfendi** [comment]
@clandry94 I think we are ready to finish the first iteration of the feature and merge it once https://github.com/Shopify/ingress/pull/120/files#r229965262 is addressed. Great work!

Once you address that last comment please revert unnecessary changes (Makefile etc.) and make an upstream PR with a link to this one.

**Review by ElvinEfendi — APPROVED** [review]
👍 Ready to make upstream PR (please use `Shopify/ingress` fork).

**Comment by ElvinEfendi** [comment]
@wayt @zrdaley can you take another look? (does not have to be this PR, you can do that in the upstream PR as well)

**Review by wayt — APPROVED** [review]
🎉

**Review by zrdaley — APPROVED** [review]
lgtm 🎉

**Comment by ElvinEfendi** [comment]
This is not part of upstream https://github.com/kubernetes/ingress-nginx/pull/3341

**Pull request closed without merging**
